File: modules/platforms/dotnet/Apache.Ignite.Core/Configuration/CacheConfiguration.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// ReSharper disable MemberCanBePrivate.Global
// ReSharper disable UnusedMember.Global
// ReSharper disable UnusedAutoPropertyAccessor.Global
namespace Apache.Ignite.Core.Configuration
{
    using Apache.Ignite.Core.Binary;
    using Apache.Ignite.Core.Cache;
    using Apache.Ignite.Core.Cache.Store;

    /// <summary>
    /// Defines grid cache configuration.
    /// </summary>
    public class CacheConfiguration
    {
        /** Default size of rebalance thread pool. */
        public const int DefaultRebalanceThreadPoolSize = 2;

        /** Default rebalance timeout (ms).*/
        public const long DefaultRebalanceTimeout = 10000;

        /** Time in milliseconds to wait between rebalance messages to avoid overloading CPU. */
        public const long DefaultRebalanceThrottle = 0;

        /** Default number of backups. */
        public const int DefaultBackups = 0;

        /** Default caching mode. */
        public const CacheMode DefaultCacheMode = CacheMode.Partitioned;

        /** Default atomicity mode. */
        public const CacheAtomicityMode DefaultCacheAtomicityMode = CacheAtomicityMode.Atomic;

        /** Default lock timeout. */
        public const long DefaultLockTimeout = 0;

        /** Initial default cache size. */
        public const int DefaultStartSize = 1500000;

        /** Default cache size to use with eviction policy. */
        public const int DefaultCacheSize = 100000;

        /** Default value for 'invalidate' flag that indicates if this is invalidation-based cache. */
        public const bool DefaultInvalidate = false;

        /** Default rebalance mode for distributed cache. */
        public const CacheRebalanceMode DefaultRebalanceMode = CacheRebalanceMode.Async;

        /** Default rebalance batch size in bytes. */
        public const int DefaultRebalanceBatchSize = 512*1024; // 512K

        /** Default maximum eviction queue ratio. */
        public const float DefaultMaxEvictionOverflowRatio = 10;

        /** Default eviction synchronized flag. */
        public const bool DefaultEvictSynchronized = false;

        /** Default eviction key buffer size for batching synchronized evicts. */
        public const int DefaultEvictKeyBufferSize = 1024;

        /** Default synchronous eviction timeout in milliseconds. */
        public const long DefaultEvictSynchronizedTimeout = 10000;

        /** Default synchronous eviction concurrency level. */
        public const int DefaultEvictSynchronizedConcurrencyLevel = 4;

        /** Default value for eager ttl flag. */
        public const bool DefaultEagerTtl = true;

        /** Default off-heap storage size is {@code -1} which means that off-heap storage is disabled. */
        public const long DefaultOffheapMemory = -1;

        /** Default value for 'swapEnabled' flag. */
        public const bool DefaultSwapEnabled = false;

        /** Default value for 'maxConcurrentAsyncOps'. */
        public const int DefaultMaxConcurrentAsyncOperations = 500;

        /** Default value for 'writeBehindEnabled' flag. */
        public const bool DefaultWriteBehindEnabled = false;

        /** Default flush size for write-behind cache store. */
        public const int DefaultWriteBehindFlushSize = 10240; // 10K

        /** Default flush frequency for write-behind cache store in milliseconds. */
        public const long DefaultWriteBehindFlushFrequency = 5000;

        /** Default count of flush threads for write-behind cache store. */
        public const int DefaultWriteFromBehindFlushThreadCount = 1;

        /** Default batch size for write-behind cache store. */
        public const int DefaultWriteBehindBatchSize = 512;

        /** Default value for load previous value flag. */
        public const bool DefaultLoadPreviousValue = false;

        /** Default memory mode. */
        public const CacheMemoryMode DefaultMemoryMode = CacheMemoryMode.OnheapTiered;

        /** Default value for 'readFromBackup' flag. */
        public const bool DefaultReadFromBackup = true;

        /** Default timeout after which long query warning will be printed. */
        public const long DefaultLongQueryWarningTimeout = 3000;

        /** Default size for onheap SQL row cache size. */
        public const int DefaultSqlOnheapRowCacheSize = 10*1024;

        /** Default value for keep portable in store behavior .*/
        public const bool DefaultKeepPortableInStore = true;

        /** Default value for 'copyOnRead' flag. */
        public const bool DefaultCopyOnRead = true;

        /// <summary>
        /// Gets or sets the cache name.
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="CacheConfiguration"/> class.
        /// </summary>
        /// <param name="name">The name.</param>
        public CacheConfiguration(string name = null)
        {
            Name = name;

            Backups = DefaultBackups;
            AtomicityMode = DefaultCacheAtomicityMode;
            CacheMode = DefaultCacheMode;
            CopyOnRead = DefaultCopyOnRead;
            StartSize = DefaultCacheSize;
            EagerTtl = DefaultEagerTtl;
            EvictSynchronizedKeyBufferSize = DefaultEvictKeyBufferSize;
            EvictSynchronized = DefaultEvictSynchronized;
            EvictSynchronizedConcurrencyLevel = DefaultEvictSynchronizedConcurrencyLevel;
            EvictSynchronizedTimeout = DefaultEvictSynchronizedTimeout;
            Invalidate = DefaultInvalidate;
            KeepPortableInStore = DefaultKeepPortableInStore;
            LoadPreviousValue = DefaultLoadPreviousValue;
            LockTimeout = DefaultLockTimeout;
            LongQueryWarningTimeout = DefaultLongQueryWarningTimeout;
            MaxConcurrentAsyncOperations = DefaultMaxConcurrentAsyncOperations;
            MaxEvictionOverflowRatio = DefaultMaxEvictionOverflowRatio;
            MemoryMode = DefaultMemoryMode;
            OffHeapMaxMemory = DefaultOffheapMemory;
            ReadFromBackup = DefaultReadFromBackup;
            RebalanceBatchSize = DefaultRebalanceBatchSize;
            RebalanceMode = DefaultRebalanceMode;
            RebalanceThreadPoolSize = DefaultRebalanceThreadPoolSize;
            RebalanceThrottle = DefaultRebalanceThrottle;
            RebalanceTimeout = DefaultRebalanceTimeout;
            SqlOnheapRowCacheSize = DefaultSqlOnheapRowCacheSize;
            StartSize = DefaultStartSize;
            StartSize = DefaultStartSize;
            EnableSwap = DefaultSwapEnabled;
            WriteBehindBatchSize = DefaultWriteBehindBatchSize;
            WriteBehindEnabled = DefaultWriteBehindEnabled;
            WriteBehindFlushFrequency = DefaultWriteBehindFlushFrequency;
            WriteBehindFlushSize = DefaultWriteBehindFlushSize;
            WriteBehindFlushThreadCount= DefaultWriteFromBehindFlushThreadCount;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="CacheConfiguration"/> class.
        /// </summary>
        /// <param name="reader">The reader.</param>
        internal CacheConfiguration(IBinaryRawReader reader)
        {
            Name = reader.ReadString();
        }

        /// <summary>
        /// Gets or sets write synchronization mode. This mode controls whether the main        
        /// caller should wait for update on other nodes to complete or not.
        /// </summary>
        public CacheWriteSynchronizationMode WriteSynchronizationMode { get; set; }

        /// <summary>
        /// Gets or sets flag indicating whether eviction is synchronized between primary, backup and near nodes.        
        /// If this parameter is true and swap is disabled then <see cref="ICache{TK,TV}.LocalEvict"/>
        /// will involve all nodes where an entry is kept.  
        /// If this property is set to false then eviction is done independently on different cache nodes.        
        /// Note that it's not recommended to set this value to true if cache store is configured since it will allow 
        /// to significantly improve cache performance.
        /// </summary>
        public bool EvictSynchronized { get; set; }

        /// <summary>
        /// Gets or sets size of the key buffer for synchronized evictions.
        /// </summary>
        public int EvictSynchronizedKeyBufferSize { get; set; }

        /// <summary>
        /// Gets or sets concurrency level for synchronized evictions. 
        /// This flag only makes sense with <see cref="EvictSynchronized"/> set to true. 
        /// When synchronized evictions are enabled, it is possible that local eviction policy will try 
        /// to evict entries faster than evictions can be synchronized with backup or near nodes. 
        /// This value specifies how many concurrent synchronous eviction sessions should be allowed 
        /// before the system is forced to wait and let synchronous evictions catch up with the eviction policy.       
        /// </summary>
        public int EvictSynchronizedConcurrencyLevel { get; set; }

        /// <summary>
        /// Gets or sets timeout for synchronized evictions
        /// </summary>
        public long EvictSynchronizedTimeout { get; set; }

        /// <summary>
        /// This value denotes the maximum size of eviction queue in percents of cache size 
        /// in case of distributed cache (replicated and partitioned) and using synchronized eviction
        /// <para/>        
        /// That queue is used internally as a buffer to decrease network costs for synchronized eviction. 
        /// Once queue size reaches specified value all required requests for all entries in the queue 
        /// are sent to remote nodes and the queue is cleared.
        /// </summary>
        public float MaxEvictionOverflowRatio { get; set; }

        /// <summary>
        /// Gets or sets flag indicating whether expired cache entries will be eagerly removed from cache. 
        /// When set to false, expired entries will be removed on next entry access.        
        /// </summary>
        public bool EagerTtl { get; set; }

        /// <summary>
        /// Gets or sets initial cache size which will be used to pre-create internal hash table after start.
        /// </summary>
        public int StartSize { get; set; }

        /// <summary>
        /// Gets or sets flag indicating whether value should be loaded from store if it is not in the cache 
        /// for the following cache operations:   
        /// <list type="bullet">
        /// <item><term><see cref="ICache{TK,TV}.PutIfAbsent"/></term></item>
        /// <item><term><see cref="ICache{TK,TV}.Replace(TK,TV)"/></term></item>
        /// <item><term><see cref="ICache{TK,TV}.Remove(TK)"/></term></item>
        /// <item><term><see cref="ICache{TK,TV}.GetAndPut"/></term></item>
        /// <item><term><see cref="ICache{TK,TV}.GetAndRemove"/></term></item>
        /// <item><term><see cref="ICache{TK,TV}.GetAndReplace"/></term></item>
        /// <item><term><see cref="ICache{TK,TV}.GetAndPutIfAbsent"/></term></item>
        /// </list>     
        /// </summary>
        public bool LoadPreviousValue { get; set; }

        /// <summary>
        /// Gets or sets factory for underlying persistent storage for read-through and write-through operations.
        /// </summary>
        public bool KeepPortableInStore { get; set; }

        /// <summary>
        /// Gets or sets caching mode to use.
        /// </summary>
        public CacheMode CacheMode { get; set; }

        /// <summary>
        /// Gets or sets cache atomicity mode.
        /// </summary>
        public CacheAtomicityMode AtomicityMode { get; set; }

        /// <summary>
        /// Gets or sets cache write ordering mode.
        /// </summary>
        public CacheAtomicWriteOrderMode AtomicWriteOrderMode { get; set; }

        /// <summary>
        /// Gets or sets number of nodes used to back up single partition for 
        /// <see cref="Configuration.CacheMode.Partitioned"/> cache.
        /// </summary>
        public int Backups { get; set; }

        /// <summary>
        /// Gets or sets default lock acquisition timeout.
        /// </summary>
        public long LockTimeout { get; set; }  // TODO: Timespan

        /// <summary>
        /// Invalidation flag. If true, values will be invalidated (nullified) upon commit in near cache.
        /// </summary>
        public bool Invalidate { get; set; }

        /// <summary>
        /// Gets or sets cache rebalance mode.
        /// </summary>
        public CacheRebalanceMode RebalanceMode { get; set; }

        /// <summary>
        /// Gets or sets size (in number bytes) to be loaded within a single rebalance message.
        /// Rebalancing algorithm will split total data set on every node into multiple batches prior to sending data.
        /// </summary>
        public int RebalanceBatchSize { get; set; }

        /// <summary>
        /// Flag indicating whether Ignite should use swap storage by default.
        /// </summary>
        public bool EnableSwap { get; set; }

        /// <summary>
        /// Gets or sets maximum number of allowed concurrent asynchronous operations, 0 for unlimited.
        /// </summary>
        public int MaxConcurrentAsyncOperations { get; set; }

        /// <summary>
        /// Flag indicating whether Ignite should use write-behind behaviour for the cache store.
        /// </summary>
        public bool WriteBehindEnabled { get; set; }

        /// <summary>
        /// Maximum size of the write-behind cache. If cache size exceeds this value, all cached items are flushed 
        /// to the cache store and write cache is cleared.
        /// </summary>
        public int WriteBehindFlushSize { get; set; }

        /// <summary>
        /// Frequency with which write-behind cache is flushed to the cache store in milliseconds.
        /// This value defines the maximum time interval between object insertion/deletion from the cache
        /// at the moment when corresponding operation is applied to the cache store.
        /// <para/> 
        /// If this value is 0, then flush is performed according to the flush size.
        /// <para/>
        /// Note that you cannot set both
        /// <see cref="WriteBehindFlushSize"/> and <see cref="WriteBehindFlushFrequency"/> to 0.
        /// </summary>
        public long WriteBehindFlushFrequency { get; set; }

        /// <summary>
        /// Number of threads that will perform cache flushing. Cache flushing is performed when cache size exceeds 
        /// value defined by <see cref="WriteBehindFlushSize"/>, or flush interval defined by 
        /// <see cref="WriteBehindFlushFrequency"/> is elapsed.
        /// </summary>
        public int WriteBehindFlushThreadCount { get; set; }

        /// <summary>
        /// Maximum batch size for write-behind cache store operations. 
        /// Store operations (get or remove) are combined in a batch of this size to be passed to 
        /// <see cref="ICacheStore.WriteAll"/> or <see cref="ICacheStore.DeleteAll"/> methods. 
        /// </summary>
        public int WriteBehindBatchSize { get; set; }

        /// <summary>
        /// Gets or sets size of rebalancing thread pool. Note that size serves as a hint and implementation 
        /// may create more threads for rebalancing than specified here (but never less threads).
        /// </summary>
        public int RebalanceThreadPoolSize { get; set; }

        /// <summary>
        /// Gets or sets rebalance timeout (ms).
        /// </summary>
        public long RebalanceTimeout { get; set; }  // TODO: Timespan

        /// <summary>
        /// Gets or sets delay in milliseconds upon a node joining or leaving topology (or crash) 
        /// after which rebalancing should be started automatically. 
        /// Rebalancing should be delayed if you plan to restart nodes
        /// after they leave topology, or if you plan to start multiple nodes at once or one after another
        /// and don't want to repartition and rebalance until all nodes are started.
        /// </summary>
        public long RebalanceDelay { get; set; }

        /// <summary>
        /// Time in milliseconds to wait between rebalance messages to avoid overloading of CPU or network.
        /// When rebalancing large data sets, the CPU or network can get over-consumed with rebalancing messages,
        /// which consecutively may slow down the application performance. This parameter helps tune 
        /// the amount of time to wait between rebalance messages to make sure that rebalancing process
        /// does not have any negative performance impact. Note that application will continue to work
        /// properly while rebalancing is still in progress.
        /// <para/>
        /// Value of 0 means that throttling is disabled.
        /// </summary>
        public long RebalanceThrottle { get; set; }

        /// <summary>
        /// Gets or sets maximum amount of memory available to off-heap storage. Possible values are
        /// -1 means that off-heap storage is disabled. 0 means that Ignite will not limit off-heap storage 
        /// (it's up to user to properly add and remove entries from cache to ensure that off-heap storage 
        /// does not grow indefinitely.
        /// Any positive value specifies the limit of off-heap storage in bytes.
        /// </summary>
        public long OffHeapMaxMemory { get; set; }

        /// <summary>
        /// Gets or sets memory mode for cache.
        /// </summary>
        public CacheMemoryMode MemoryMode { get; set; }

        /// <summary>
        /// Gets or sets flag indicating whether data can be read from backup.
        /// </summary>
        public bool ReadFromBackup { get; set; }

        /// <summary>
        /// Gets or sets flag indicating whether copy of of the value stored in cache should be created
        /// for cache operation implying return value. 
        /// </summary>
        public bool CopyOnRead { get; set; }

        /// <summary>
        /// Gets or sets the timeout after which long query warning will be printed.
        /// </summary>
        public long LongQueryWarningTimeout { get; set; }  // TODO

        /// <summary>
        /// If true all the SQL table and field names will be escaped with double quotes like 
        /// ({ "tableName"."fieldsName"}). This enforces case sensitivity for field names and
        /// also allows having special characters in table and field names.
        /// </summary>
        public bool SqlEscapeAll { get; set; }

        /// <summary>
        /// Number of SQL rows which will be cached onheap to avoid deserialization on each SQL index access.
        /// This setting only makes sense when offheap is enabled for this cache.
        /// </summary>
        public int SqlOnheapRowCacheSize { get; set; }
    }
}

File: modules/core/src/main/java/org/apache/ignite/internal/management/cache/CacheConfiguration.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.management.cache;

import java.util.List;
import org.apache.ignite.cache.CacheAtomicityMode;
import org.apache.ignite.cache.CacheMode;
import org.apache.ignite.cache.CacheWriteSynchronizationMode;
import org.apache.ignite.cache.PartitionLossPolicy;
import org.apache.ignite.configuration.DataRegionConfiguration;
import org.apache.ignite.configuration.DiskPageCompression;
import org.apache.ignite.internal.IgniteEx;
import org.apache.ignite.internal.Order;
import org.apache.ignite.internal.dto.IgniteDataTransferObject;
import org.apache.ignite.internal.util.typedef.internal.S;
import org.apache.ignite.lang.IgniteUuid;
import org.jetbrains.annotations.Nullable;

import static org.apache.ignite.internal.visor.util.VisorTaskUtils.compactClass;
import static org.apache.ignite.internal.visor.util.VisorTaskUtils.compactIterable;

/**
 * Data transfer object for cache configuration properties.
 */
public class CacheConfiguration extends IgniteDataTransferObject {
    /** */
    private static final long serialVersionUID = 0L;

    /** Cache name. */
    @Order(0)
    String name;

    /** Cache group name. */
    @Order(1)
    String grpName;

    /** Cache mode. */
    @Order(2)
    CacheMode mode;

    /** Cache atomicity mode. */
    @Order(3)
    CacheAtomicityMode atomicityMode;

    /** Eager ttl flag. */
    @Order(4)
    boolean eagerTtl;

    /** Write synchronization mode. */
    @Order(5)
    CacheWriteSynchronizationMode writeSynchronizationMode;

    /** Invalidate. */
    @Order(6)
    boolean invalidate;

    /** Max concurrent async operations. */
    @Order(7)
    int maxConcurrentAsyncOps;

    /** Cache interceptor. */
    @Order(8)
    String interceptor;

    /** Default lock acquisition timeout. */
    @Order(9)
    long dfltLockTimeout;

    /** Cache affinity config. */
    @Order(10)
    CacheAffinityConfiguration affinityCfg;

    /** Preload config. */
    @Order(11)
    CacheRebalanceConfiguration rebalanceCfg;

    /** Eviction config. */
    @Order(12)
    CacheEvictionConfiguration evictCfg;

    /** Near cache config. */
    @Order(13)
    CacheNearConfiguration nearCfg;

    /** Store config. */
    @Order(14)
    CacheStoreConfiguration storeCfg;

    /** Collection of query entities. */
    @Order(15)
    List<QueryEntity> qryEntities;

    /** Collection of type metadata. */
    @Order(16)
    List<CacheJdbcType> jdbcTypes;

    /** Whether statistics collection is enabled. */
    @Order(17)
    boolean statisticsEnabled;

    /** Whether management is enabled. */
    @Order(18)
    boolean mgmtEnabled;

    /** Class name of cache loader factory. */
    @Order(19)
    String ldrFactory;

    /** Class name of cache writer factory. */
    @Order(20)
    String writerFactory;

    /** Class name of expiry policy factory. */
    @Order(21)
    String expiryPlcFactory;

    /** Query configuration. */
    @Order(22)
    QueryConfiguration qryCfg;

    /** System cache flag. */
    @Order(23)
    boolean sys;

    /** Keep binary in store flag. */
    @Order(24)
    boolean storeKeepBinary;

    /** On-heap cache enabled flag. */
    @Order(25)
    boolean onheapCache;

    /** Partition loss policy. */
    @Order(26)
    PartitionLossPolicy partLossPlc;

    /** Query parallelism. */
    @Order(27)
    int qryParallelism;

    /** Copy on read flag. */
    @Order(28)
    boolean cpOnRead;

    /** Eviction filter. */
    @Order(29)
    String evictFilter;

    /** Listener configurations. */
    @Order(30)
    String lsnrConfigurations;

    /** */
    @Order(31)
    boolean loadPrevVal;

    /** Name of {@link DataRegionConfiguration} for this cache */
    @Order(32)
    String dataRegName;

    /** Maximum inline size for sql indexes. */
    @Order(33)
    int sqlIdxMaxInlineSize;

    /** Node filter specifying nodes on which this cache should be deployed. */
    @Order(34)
    String nodeFilter;

    /** */
    @Order(35)
    int qryDetailMetricsSz;

    /** Flag indicating whether data can be read from backup. */
    @Order(36)
    boolean readFromBackup;

    /** Name of class implementing GridCacheTmLookup. */
    @Order(37)
    String tmLookupClsName;

    /** Cache topology validator. */
    @Order(38)
    String topValidator;

    /** Dynamic deployment ID. */
    @Order(39)
    IgniteUuid dynamicDeploymentId;

    /** Disk page compression algorithm. */
    @Order(40)
    DiskPageCompression diskPageCompression;

    /** Algorithm specific disk page compression level. */
    @Order(41)
    Integer diskPageCompressionLevel;

    /**
     * Default constructor.
     */
    public CacheConfiguration() {
        // No-op.
    }

    /**
     * Create data transfer object for cache configuration properties.
     *
     * @param ignite Grid.
     * @param ccfg Cache configuration.
     * @param dynamicDeploymentId Dynamic deployment ID.
     */
    public CacheConfiguration(IgniteEx ignite, org.apache.ignite.configuration.CacheConfiguration ccfg, IgniteUuid dynamicDeploymentId) {
        name = ccfg.getName();
        grpName = ccfg.getGroupName();
        this.dynamicDeploymentId = dynamicDeploymentId;
        mode = ccfg.getCacheMode();
        atomicityMode = ccfg.getAtomicityMode();
        eagerTtl = ccfg.isEagerTtl();
        writeSynchronizationMode = ccfg.getWriteSynchronizationMode();
        invalidate = ccfg.isInvalidate();
        maxConcurrentAsyncOps = ccfg.getMaxConcurrentAsyncOperations();
        interceptor = compactClass(ccfg.getInterceptor());
        dfltLockTimeout = ccfg.getDefaultLockTimeout();
        qryEntities = QueryEntity.list(ccfg.getQueryEntities());
        jdbcTypes = CacheJdbcType.list(ccfg.getCacheStoreFactory());
        statisticsEnabled = ccfg.isStatisticsEnabled();
        mgmtEnabled = ccfg.isManagementEnabled();
        ldrFactory = compactClass(ccfg.getCacheLoaderFactory());
        writerFactory = compactClass(ccfg.getCacheWriterFactory());
        expiryPlcFactory = compactClass(ccfg.getExpiryPolicyFactory());

        sys = ignite.context().cache().systemCache(ccfg.getName());
        storeKeepBinary = ccfg.isStoreKeepBinary();
        onheapCache = ccfg.isOnheapCacheEnabled();
        partLossPlc = ccfg.getPartitionLossPolicy();
        qryParallelism = ccfg.getQueryParallelism();

        affinityCfg = new CacheAffinityConfiguration(ccfg);
        rebalanceCfg = new CacheRebalanceConfiguration(ccfg);
        evictCfg = new CacheEvictionConfiguration(ccfg);
        nearCfg = new CacheNearConfiguration(ccfg);

        storeCfg = new CacheStoreConfiguration(ignite, ccfg);

        qryCfg = new QueryConfiguration(ccfg);

        cpOnRead = ccfg.isCopyOnRead();
        evictFilter = compactClass(ccfg.getEvictionFilter());
        lsnrConfigurations = compactIterable(ccfg.getCacheEntryListenerConfigurations());
        loadPrevVal = ccfg.isLoadPreviousValue();
        dataRegName = ccfg.getDataRegionName();
        sqlIdxMaxInlineSize = ccfg.getSqlIndexMaxInlineSize();
        nodeFilter = compactClass(ccfg.getNodeFilter());
        qryDetailMetricsSz = ccfg.getQueryDetailMetricsSize();
        readFromBackup = ccfg.isReadFromBackup();
        tmLookupClsName = ccfg.getTransactionManagerLookupClassName();
        topValidator = compactClass(ccfg.getTopologyValidator());

        diskPageCompression = ccfg.getDiskPageCompression();
        diskPageCompressionLevel = ccfg.getDiskPageCompressionLevel();
    }

    /**
     * @return Cache name.
     */
    @Nullable public String getName() {
        return name;
    }

    /**
     * @return Cache group name.
     */
    @Nullable public String getGroupName() {
        return grpName;
    }

    /**
     * @return Cache mode.
     */
    public CacheMode getMode() {
        return mode;
    }

    /**
     * @return Cache atomicity mode.
     */
    public CacheAtomicityMode getAtomicityMode() {
        return atomicityMode;
    }

    /**
     * @return Eager ttl flag
     */
    public boolean eagerTtl() {
        return eagerTtl;
    }

    /**
     * @return Write synchronization mode.
     */
    public CacheWriteSynchronizationMode getWriteSynchronizationMode() {
        return writeSynchronizationMode;
    }

    /**
     * @return Invalidate.
     */
    public boolean isInvalidate() {
        return invalidate;
    }

    /**
     * @return Max concurrent async operations
     */
    public int getMaxConcurrentAsyncOperations() {
        return maxConcurrentAsyncOps;
    }

    /**
     * @return Cache interceptor.
     */
    @Nullable public String getInterceptor() {
        return interceptor;
    }

    /**
     * @return Gets default lock acquisition timeout.
     */
    public long getDefaultLockTimeout() {
        return dfltLockTimeout;
    }

    /**
     * @return Collection of type metadata.
     */
    public List<CacheJdbcType> getJdbcTypes() {
        return jdbcTypes;
    }

    /**
     * @return Near cache config.
     */
    public CacheNearConfiguration getNearConfiguration() {
        return nearCfg;
    }

    /**
     * @return Eager ttl flag.
     */
    public boolean isEagerTtl() {
        return eagerTtl;
    }

    /**
     * @return {@code true} if cache statistics collection enabled.
     */
    public boolean isStatisticsEnabled() {
        return statisticsEnabled;
    }

    /**
     * @return Whether management is enabled.
     */
    public boolean isManagementEnabled() {
        return mgmtEnabled;
    }

    /**
     * @return Class name of cache loader factory.
     */
    public String getLoaderFactory() {
        return ldrFactory;
    }

    /**
     * @return Class name of cache writer factory.
     */
    public String getWriterFactory() {
        return writerFactory;
    }

    /**
     * @return Class name of expiry policy factory.
     */
    public String getExpiryPolicyFactory() {
        return expiryPlcFactory;
    }

    /**
     * @return Cache affinity config.
     */
    public CacheAffinityConfiguration getAffinityConfiguration() {
        return affinityCfg;
    }

    /**
     * @return Preload config.
     */
    public CacheRebalanceConfiguration getRebalanceConfiguration() {
        return rebalanceCfg;
    }

    /**
     * @return Eviction config.
     */
    public CacheEvictionConfiguration getEvictionConfiguration() {
        return evictCfg;
    }

    /**
     * @return Store config
     */
    public CacheStoreConfiguration getStoreConfiguration() {
        return storeCfg;
    }

    /**
     * @return Collection of query entities.
     */
    public List<QueryEntity> getQueryEntities() {
        return qryEntities;
    }

    /**
     * @return Collection of query entities.
     */
    public QueryConfiguration getQueryConfiguration() {
        return qryCfg;
    }

    /**
     * @return System cache flag.
     */
    public boolean isSystem() {
        return sys;
    }

    /**
     * @return Keep binary in store flag.
     */
    public Boolean isStoreKeepBinary() {
        return storeKeepBinary;
    }

    /**
     * @return On-heap cache enabled flag.
     */
    public boolean isOnheapCacheEnabled() {
        return onheapCache;
    }

    /**
     * @return Partition loss policy.
     */
    public PartitionLossPolicy getPartitionLossPolicy() {
        return partLossPlc;
    }

    /**
     * @return Query parallelism.
     */
    public int getQueryParallelism() {
        return qryParallelism;
    }

    /**
     * @return Copy on read flag.
     */
    public boolean isCopyOnRead() {
        return cpOnRead;
    }

    /**
     * @return Eviction filter or {@code null}.
     */
    public String getEvictionFilter() {
        return evictFilter;
    }

    /**
     * @return Listener configurations.
     */
    public String getListenerConfigurations() {
        return lsnrConfigurations;
    }

    /**
     * @return Load previous value flag.
     */
    public boolean isLoadPreviousValue() {
        return loadPrevVal;
    }

    /**
     * @return {@link DataRegionConfiguration} name.
     */
    @Deprecated
    public String getMemoryPolicyName() {
        return dataRegName;
    }

    /**
     * @return Maximum payload size for offheap indexes.
     */
    public int getSqlIndexMaxInlineSize() {
        return sqlIdxMaxInlineSize;
    }

    /**
     * @return Predicate specifying on which nodes the cache should be started.
     */
    public String getNodeFilter() {
        return nodeFilter;
    }

    /**
     * @return Maximum number of query metrics that will be stored in memory.
     */
    public int getQueryDetailMetricsSize() {
        return qryDetailMetricsSz;
    }

    /**
     * @return {@code true} if data can be read from backup node or {@code false} if data always should be read from
     * primary node and never from backup.
     */
    public boolean isReadFromBackup() {
        return readFromBackup;
    }

    /**
     * @return Transaction manager finder.
     */
    @Deprecated
    public String getTransactionManagerLookupClassName() {
        return tmLookupClsName;
    }

    /**
     * @return Topology validator.
     */
    public String getTopologyValidator() {
        return topValidator;
    }

    /**
     * @return Cache dynamic deployment ID.
     */
    public IgniteUuid getDynamicDeploymentId() {
        return dynamicDeploymentId;
    }

    /**
     * @return Disk page compression algorithm.
     */
    public DiskPageCompression getDiskPageCompression() {
        return diskPageCompression;
    }

    /**
     * @return Algorithm specific disk page compression level.
     */
    public Integer getDiskPageCompressionLevel() {
        return diskPageCompressionLevel;
    }

    /** {@inheritDoc} */
    @Override public String toString() {
        return S.toString(CacheConfiguration.class, this);
    }
}
